typing_inspect.py

```diff
@@ -435,3 +543,177 @@ class Other(dict):
+            else:
+                try:
+                    cls_args = cls.__args__
+                except AttributeError:
+                    cls_args = ()
+            return cls_args if cls_args is not None else ()
+
+        for arg in _get_args(cls):
+            tree_args.append(_replace_arg(arg, tvars, args))
+        # ... then continue replacing down the origin chain.
+        for ocls in orig_chain:
+            new_tree_args = []
+            for arg in _get_args(ocls):
+                new_tree_args.append(_replace_arg(arg, get_parameters(ocls), tree_args))
+            tree_args = new_tree_args
+        return tree_args
+
+    def _union_subs_tree(tp, tvars=None, args=None):
+        """ backport of Union._subs_tree """
+        if tp is Union:
+            return Union  # Nothing to substitute
+        tree_args = _subs_tree(tp, tvars, args)
+        # tree_args = tp.__union_params__ if tp.__union_params__ is not None else ()
+        tree_args = _remove_dups_flatten(tree_args)
+        if len(tree_args) == 1:
+            return tree_args[0]  # Union of a single type is that type
+        return (Union,) + tree_args
+
+    def _generic_subs_tree(tp, tvars=None, args=None):
+        """ backport of GenericMeta._subs_tree """
+        if tp.__origin__ is None:
+            return tp
+        tree_args = _subs_tree(tp, tvars, args)
+        return (_gorg(tp),) + tuple(tree_args)
+
+    def _tuple_subs_tree(tp, tvars=None, args=None):
+        """ ad-hoc function (inspired by union) for legacy typing """
+        if tp is Tuple:
+            return Tuple  # Nothing to substitute
+        tree_args = _subs_tree(tp, tvars, args)
+        return (Tuple,) + tuple(tree_args)
+
+    def _has_type_var(t):
+        if t is None:
+            return False
+        elif is_union_type(t):
+            return _union_has_type_var(t)
+        elif is_tuple_type(t):
+            return _tuple_has_type_var(t)
+        elif is_generic_type(t):
+            return _generic_has_type_var(t)
+        elif is_callable_type(t):
+            return _callable_has_type_var(t)
+        else:
+            return False
+
+    def _union_has_type_var(tp):
+        if tp.__union_params__:
+            for t in tp.__union_params__:
+                if _has_type_var(t):
+                    return True
+        return False
+
+    def _tuple_has_type_var(tp):
+        if tp.__tuple_params__:
+            for t in tp.__tuple_params__:
+                if _has_type_var(t):
+                    return True
+        return False
+
+    def _callable_has_type_var(tp):
+        if tp.__args__:
+            for t in tp.__args__:
+                if _has_type_var(t):
+                    return True
+        return _has_type_var(tp.__result__)
+
+    def _generic_has_type_var(tp):
+        if tp.__parameters__:
+            for t in tp.__parameters__:
+                if _has_type_var(t):
+                    return True
+        return False
```

Comment:
No need to put the private helpers under if. It is fine to always define them.

Reply:
Again an optimization here. But as you wish, I remove the if